web/app.js

```diff
@@ -2674,16 +2694,51 @@ function webViewerWheel(evt) {
 
 function webViewerTouchStart(evt) {
   if (evt.touches.length > 1) {
-    // Disable touch-based zooming, because the entire UI bits gets zoomed and
-    // that doesn't look great. If we do want to have a good touch-based
-    // zooming experience, we need to implement smooth zoom capability (probably
-    // using a CSS transform for faster visual response, followed by async
-    // re-rendering at the final zoom level) and do gesture detection on the
-    // touchmove events to drive it. Or if we want to settle for a less good
-    // experience we can make the touchmove events drive the existing step-zoom
-    // behaviour that the ctrl+mousewheel path takes.
     evt.preventDefault();
+
+    if (evt.touches.length !== 2) {
+      PDFViewerApplication._touchInfo = null;
+      return;
+    }
+
+    const [touch0, touch1] = evt.touches;
+    PDFViewerApplication._touchInfo = {
+      centerX: (touch0.pageX + touch1.pageX) / 2,
+      centerY: (touch0.pageY + touch1.pageY) / 2,
+      distance:
+        Math.hypot(touch0.pageX - touch1.pageX, touch0.pageY - touch1.pageY) ||
+        1,
+    };
+  }
+}
+
+function webViewerTouchMove(evt) {
+  if (!PDFViewerApplication._touchInfo || evt.touches.length !== 2) {
+    return;
+  }
+
+  evt.preventDefault();
+  const [touch0, touch1] = evt.touches;
+  const distance =
+    Math.hypot(touch0.pageX - touch1.pageX, touch0.pageY - touch1.pageY) || 1;
+  const scaleFactor = distance / PDFViewerApplication._touchInfo.distance;
```

Comment:
*A more general question:* Do we actually need to support "arbitrary" pinch-zooming, or could we perhaps convert the distance changes into zoom-`ticks` similar to how the mouse-wheel zooming is implemented?

Unless I'm missing something obvious, it seems to me that that could actually simplify the overall implementation.

Reply:
I tested zooming on an android device with the Acrobat app and the Google Drive app and both are supporting "smooth" zooming.
In Firefox, there's a smooth zoom when `apz.allow_zooming` is true and a stepped one else (and I forgot to check that in this patch, I'll update).